File: src/main/java/brightspark/runicmagic/init/RMItems.java
package brightspark.runicmagic.init;

import brightspark.runicmagic.RunicMagic;
import brightspark.runicmagic.enums.StaffType;
import brightspark.runicmagic.item.*;
import net.minecraft.item.Item;
import net.minecraftforge.fml.common.registry.GameRegistry;

import java.util.LinkedList;
import java.util.List;

@GameRegistry.ObjectHolder(RunicMagic.MOD_ID)
public class RMItems
{
	private static List<Item> ITEMS = new LinkedList<>();

	//Crafting Ingredients
	public static final Item rune_essence = null;
	public static final Item pure_essence = null;
	public static final Item rune = null;
	public static final Item talisman = null;
	public static final Item orb = null;

	//Staffs
	public static final Item staff_basic = null;
	public static final Item staff_battle = null;
	public static final Item staff_mystic = null;
	//public static final Item staff_limitless = null;

	private static void add(Item item)
	{
		ITEMS.add(item);
	}

	public static void init()
	{
		add(new RMItemBase("rune_essence"));
		add(new RMItemBase("pure_essence"));
		add(new ItemRune());
        add(new ItemTalisman());
		add(new RMItemSubBase("orb", "unpowered", "air", "water", "earth", "fire").setMaxStackSize(1));

		for(StaffType type : StaffType.values())
			add(new ItemStaff(type));
	}

	public static Item[] getItems()
	{
		if(ITEMS.isEmpty())
			init();
		return ITEMS.toArray(new Item[0]);
	}
}

File: src/main/kotlin/brightspark/runicmagic/init/RMItems.kt
package brightspark.runicmagic.init

import brightspark.runicmagic.RunicMagic
import brightspark.runicmagic.item.RuneItem
import brightspark.runicmagic.item.StaffItem
import brightspark.runicmagic.item.TalismanItem
import brightspark.runicmagic.model.RuneType
import brightspark.runicmagic.model.StaffType
import brightspark.runicmagic.util.setRegName
import net.minecraft.block.Block
import net.minecraft.item.BlockItem
import net.minecraft.item.Item
import net.minecraftforge.event.RegistryEvent
import thedarkcolour.kotlinforforge.forge.objectHolder

object RMItems {
	val RUNE_ESSENCE: Item by objectHolder("rune_essence")
	val PURE_ESSENCE: Item by objectHolder("pure_essence")

	val ORB_NONE: Item by objectHolder("orb_none")
	val ORB_AIR: Item by objectHolder("orb_air")
	val ORB_WATER: Item by objectHolder("orb_water")
	val ORB_EARTH: Item by objectHolder("orb_earth")
	val ORB_FIRE: Item by objectHolder("orb_fire")

	fun register(event: RegistryEvent.Register<Item>) = event.registry.registerAll(
		// Essences
		item("rune_essence"),
		item("pure_essence"),
		// Runes
		*RuneType.values().filter { it != RuneType.NONE }.map { runeItem("rune_$it", it) }.toTypedArray(),
		// Talismans
		*RuneType.values().filter { it.hasTalisman }.map { talismanItem(it) }.toTypedArray(),
		// Staves
		*StaffType.values().flatMap { staffType -> staffType.runeTypes.map { staffItem(staffType, it) } }.toTypedArray(),
		// Orbs
		runeItem("orb_none", RuneType.NONE, props().maxStackSize(1)),
		*RuneType.ELEMENTAL_TYPES.map { runeItem("orb_$it", it, props().maxStackSize(1)) }.toTypedArray(),

		// Item blocks
		*RMBlocks.BLOCKS.map { blockItem(it) }.toTypedArray()
	)

	private fun props(): Item.Properties = Item.Properties().apply { group(RunicMagic.GROUP) }

	private fun item(name: String, props: Item.Properties = props()): Item = Item(props).setRegName(name)

	private fun runeItem(name: String, runeType: RuneType, props: Item.Properties = props()): Item =
		RuneItem(runeType, props).setRegName(name)

	private fun talismanItem(runeType: RuneType): Item =
		TalismanItem(runeType, props().maxStackSize(1)).setRegName("talisman_$runeType")

	private fun staffItem(staffType: StaffType, runeType: RuneType): Item =
		StaffItem(staffType, runeType, props().maxStackSize(1)).setRegName("staff_${staffType}_$runeType")

	private fun blockItem(block: Block): Item =
		BlockItem(block, props()).setRegName(block.registryName!!.path)
}
